File: python/trees/Solution1448.py
# 1448. 统计二叉树中好节点的数目
from typing import Optional

from trees import TreeNode


class Solution:
    def goodNodes(self, root: Optional[TreeNode]) -> int:

        if root is None:
            return 0

        self.num = 0

        def dfs(node, max_value):
            if node is None:
                return
            value = node.val
            if value >= max_value:
                self.num += 1
            # dfs
            dfs(node.left, max(value, max_value))
            dfs(node.right, max(value, max_value))
            return

        dfs(root, root.val - 1)
        return self.num

File: src/tree/Solution1448.java
// 1448. 统计二叉树中好节点的数目
package tree;

public class Solution1448 {

    int num = 0;

    public int goodNodes(TreeNode root) {
        dfs(root, root.val - 1);
        return num;
    }

    public void dfs(TreeNode node, int maxValue) {
        if (node == null) return;
        if (node.val >= maxValue) num++;
        dfs(node.left, Math.max(maxValue, node.val));
        dfs(node.right, Math.max(maxValue, node.val));
    }

}
